doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -11,92 +11,83 @@ These tactics use the equality :g:`eq:forall A:Type, A->A->Prop` defined in
 
-   .. tacv:: rewrite @term at @occurrences
+   :n:`{? {| -> | <- } }`
+     For `->` (the default), :n:`term__1` is rewritten
+     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term_1`.
 
-      Rewrite only the given :token:`occurrences` of :token:`term`. Occurrences are
-      specified from left to right as for pattern (:tacn:`pattern`). The rewrite is
-      always performed using setoid rewriting, even for Leibniz’s equality, so one
-      has to ``Import Setoid`` to use this variant.
+   :n:`@natural`
+     If followed by :n:`?`, gives the maximum number of rewrites that will be performed.
+     Otherwise, it specifies the exact number of rewrites to perform (whether followed by
+     `!` or not).  If the given number of rewrites can't be done, the tactic fails.
 
-   .. tacv:: rewrite @term by @tactic
+   :n:`?`
+     Performs the rewrite as many times as possible (possibly zero times). This form never fails.
 
-      Use tactic to completely solve the side-conditions arising from the
-      :tacn:`rewrite`.
+   :n:`!`
+     Performs the rewrite as many times as possible and at least once.  Otherwise this
+     form fails.
 
-   .. tacv:: rewrite {+, @orientation @term} {? in @ident }
+   .. todo PR: The following seems to be standard behavior for @occurrences, so I plan
+      to drop (most of?) it.
+      And I don't get the part about
+      "for all hypotheses :g:`H'` different from :g:`H`".  What do these mean?
 
-      Is equivalent to the `n` successive tactics :n:`{+; rewrite @term}`, each one
-      working on the first subgoal generated by the previous one. An :production:`orientation`
-      ``->`` or ``<-`` can be inserted before each :token:`term` to rewrite.  One
-      unique clause can be added at the end after the keyword in; it will then
-      affect all rewrite operations.
+   + :n:`rewrite T in * |-` will do :n:`rewrite T in H` for all hypotheses
+     :g:`H` different from :g:`T`.
+     A success will happen as soon as at least one of these simpler tactics succeeds.
 
-   In all forms of rewrite described above, a :token:`term` to rewrite can be
-   immediately prefixed by one of the following modifiers:
+   .. exn:: The @term provided does not end with an equation.
+      :undocumented:
 
-   + `?` : the tactic :n:`rewrite ?@term` performs the rewrite of :token:`term` as many
-     times as possible (perhaps zero time). This form never fails.
-   + :n:`@natural?` : works similarly, except that it will do at most :token:`natural` rewrites.
-   + `!` : works as `?`, except that at least one rewrite should succeed, otherwise
-     the tactic fails.
-   + :n:`@natural!` (or simply :n:`@natural`) : precisely :token:`natural` rewrites of :token:`term` will be done,
-     leading to failure if these :token:`natural` rewrites are not possible.
+   .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.
```

Comment:
It looked like the second sentence was actually the description of the error message, not a part of it. But it is also an incorrect description AFAICT. The error message can occur in this case:

```coq
Parameter ax : 0 = 0.

Goal 0 = 1.
rewrite ax.
```

But when `term` does not occur in the goal, the usual error message is rather: "Found no subterm matching "..." in the current goal."